### DISCUSSION TITLE
Using specific memory regions in MicroPython modules (DTCM, etc.)

### DISCUSSION THREAD

**Discussion by graeme-winter** [micropython/micropython#14367] [m.iMX RT / Teensy 4.x]
Default configuration of MicroPython for Teensy 4.1 has 768kB of OCRAM2+OCRAM as the heap, which makes sense for the general use case. If however I want to explicitly use some DTCM or the OCRAM region that is in the TCM, I don't have an option to `m_malloc` in that region in my extension module or define an `array` in the same area?

Key part of linker output:

```
m_itcm           0x0000000000000000 0x0000000000020000 xr
m_dtcm           0x0000000020000000 0x0000000000020000 rw
m_ocrm           0x0000000020200000 0x00000000000c0000 rw
```

From what I can see on a quick look the TCM isn't really used at all at least explicitly, and it is also true that the relatively large caches in the CPU would mitigate most of the performance issues, but I was still curious as to how this space could be used.

Specifics:
- when loading a module, it would be good to put the text into `m_itcm`?
- if I have a performance-critical buffer could I put it into the TCM region of the `m_ocrm`?

It would also look like the vector table is stored in flash - again maybe this would be cached so wouldn't matter?

```
>>> from machine import mem32
>>> hex(mem32[0xE000ED08])
'0x60002000'
```

which aligns with 

```
m_flash_config   0x0000000060000000 0x0000000000001000 xr
m_ivt            0x0000000060001000 0x0000000000001000 xr
m_interrupts     0x0000000060002000 0x0000000000000400 xr
m_text           0x0000000060002400 0x00000000000fdc00 xr
```

I noticed when trying to copy the vector table and update VTOR the board would lock up (I have done this on M0+ and M4 just fine) so again wonder if the cache is doing a lot of work here.

I appreciate using MicroPython on MCU (though this is not really that micro) for relatively heavy lifting is outside the usual and I could just use a proper RTOS like Zephyr, but I was curious.

**Comment by robert-hh** [comment]
> when loading a module, it would be good to put the text into m_itcm?

ITCM is already used for code. See line 100 and 205 in boards/common.ld. But the areas could be made smaller, if not used.

**Reply by graeme-winter** [reply]
Thank you - I was thinking about `import module` where this is a compiled `.mpy` file - however I think I need to better understand the way the linker scripts work & also write a module where I return pointers to the functions to make sure I understand what is going on...

The primary question of how I can allocate memory in TCM rather than OCRAM was the main one though (I should probably have split this into two discussions) though I don't have an idea of how much difference this will make to performance
